tar/write.c

```diff
@@ -284,38 +285,63 @@ tarsnap_mode_c(struct bsdtar *bsdtar)
 	    bsdtar->argc_orig, bsdtar->argv_orig,
 	    bsdtar->option_print_stats, bsdtar->option_dryrun,
 	    bsdtar->creationtime);
-	if (bsdtar->write_cookie == NULL)
-		bsdtar_errc(bsdtar, 1, 0, "%s", archive_error_string(a));
+	if (bsdtar->write_cookie == NULL) {
+		bsdtar_warnc(bsdtar, 0, "%s", archive_error_string(a));
+		goto err0;
+	}
 
 	/*
 	 * Remember the device and inode numbers of the cache directory, so
 	 * that we can skip it in write_hierarchy().
 	 */
 	if (getdevino(a, bsdtar->cachedir,
-	    &bsdtar->cachedir_dev, &bsdtar->cachedir_ino))
-		bsdtar_errc(bsdtar, 1, 0, "%s", archive_error_string(a));
+	    &bsdtar->cachedir_dev, &bsdtar->cachedir_ino)) {
+		bsdtar_warnc(bsdtar, 0, "%s", archive_error_string(a));
+		goto err0;
+	}
 
 	/* If the chunkification cache is enabled, read it. */
 	if ((bsdtar->cachecrunch < 2) && (bsdtar->cachedir != NULL)) {
 		bsdtar->chunk_cache = ccache_read(bsdtar->cachedir);
-		if (bsdtar->chunk_cache == NULL)
-			bsdtar_errc(bsdtar, 1, errno, "Error reading cache");
+		if (bsdtar->chunk_cache == NULL) {
+			bsdtar_warnc(bsdtar, errno, "Error reading cache");
+			goto err0;
+		}
 	}
 
+	/*
+	 * write_archive(a, bsdtar) calls archive_write_finish(a), so we set
+	 * a = NULL to avoid doing this again in err1.
+	 */
 	write_archive(a, bsdtar);
+	a = NULL;
 
 	/*
 	 * If this isn't a dry run and we're running with the chunkification
 	 * cache enabled, write the cache back to disk.
 	 */
 	if ((bsdtar->option_dryrun == 0) && (bsdtar->cachecrunch < 2)) {
-		if (ccache_write(bsdtar->chunk_cache, bsdtar->cachedir))
-			bsdtar_errc(bsdtar, 1, errno, "Error writing cache");
+		if (ccache_write(bsdtar->chunk_cache, bsdtar->cachedir)) {
+			bsdtar_warnc(bsdtar, errno, "Error writing cache");
+			goto err1;
+		}
 	}
 
 	/* Free the chunkification cache. */
 	if (bsdtar->cachecrunch < 2)
 		ccache_free(bsdtar->chunk_cache);
+
+	/* Success */
+	return;
+
+err1:
+	ccache_free(bsdtar->chunk_cache);
+err0:
```

Comment:
done.  I could have added an err0 for the initial malloc erroc, but since you didn't mention that, I left it as `bsdtar_errc`.  I'll obviously change that too if you'd like.

in favor of changing: that seems to match other tarsnap style better (looking at random networking and crypto files)

against changing: we've got the `bsdtar` pointer, so we might as well call `bsdtar_errc` directly.